crates/sparrow-transforms/src/transform_pipeline.rs

```diff
@@ -0,0 +1,174 @@
+use std::collections::VecDeque;
+use std::sync::Arc;
+
+use error_stack::ResultExt;
+use itertools::Itertools;
+use parking_lot::Mutex;
+use sparrow_arrow::Batch;
+use sparrow_scheduler::{
+    Partition, Partitioned, Pipeline, PipelineSink, Queue, Scheduler, Sink, TaskRef,
+};
+
+use crate::transforms::Transform;
+use crate::Error;
+
+/// Runs a linear sequence of transforms as a pipeline.
+pub struct TransformPipeline {
+    /// The inputs this transform processes.
+    ///
+    /// For each partition:
+    ///   Hold a shared, mutex-protected input queue.
+    inputs: Partitioned<Arc<Mutex<InputBuffer>>>,
+    transforms: Vec<Box<dyn Transform>>,
+    /// Compute tasks to wake up as needed.
+    tasks: Partitioned<TaskRef>,
+    /// Sink for the down-stream computation.
+    sink: PipelineSink,
+}
+
+impl std::fmt::Debug for TransformPipeline {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("TransformPipeline")
+            .field(
+                "transforms",
+                &self.transforms.iter().map(|t| t.name()).format(","),
+            )
+            .finish()
+    }
+}
+
+#[derive(Debug, Default)]
+struct InputBuffer {
+    buffer: VecDeque<Batch>,
+    is_closed: bool,
+}
+
+impl TransformPipeline {
```

Comment:
This shows an example of creating a pipeline. This one runs a sequence of 1 or more transforms on an input batch. Note that it does some "low-ish" level mutex protections. This is partly intentional. As described in `pipeline.rs` it allows us to implement a variety of workflows by changing how the pipeline chooses to be implemented:

1. It could do more work in `push` which would block the producer.
2. It can do more in `do_work`

In general, `push` and `close` do a little bit of work to figure out what to wake up and push the work there. They are (ideally) short-lived, and don't hold the partition lock for long.

Do work shouldn't hold the lock while doing the bulk of the work.